adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java

```diff
@@ -195,80 +192,354 @@ public void run() {
+                if (params.containsKey(key)) {
+                    Logger.w(TAG,
+                            "Key/value pair list contains redundant key. ",
+                            "Redundant key: " + key,
+                            ADALError.DEVELOPER_BEARER_HEADER_MULTIPLE_ITEMS);
                 }
-            } else {
+
+                // Add the key/value to the Map
+                params.put(key, value);
+            }
+
+            if (params.isEmpty()) { // To match the existing expected behavior, an Exception is thrown.
+                throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
+            }
+
+            return params;
+        }
+
+        /**
+         * Parses multiple challenges in a single String, typically the value field of a WWW-Authenticate header.
+         *
+         * @param strChallenges The challenge String to parse.
+         * @return The Challenge object derived from challenge.
+         * @throws ResourceAuthenticationChallengeException If a parsing error is encountered or
+         *                                                  the String is malformed.
+         */
+        static List<Challenge> parseChallenges(final String strChallenges) throws ResourceAuthenticationChallengeException {
+            // Initialize and out-List for our result.
+            final List<Challenge> challenges = new ArrayList<>();
+
+            try { // Separate the challenges.
+                List<String> strChallengesList = separateChallenges(strChallenges);
+
+                // Add each to the out-List
+                for (final String challenge : strChallengesList) {
+                    challenges.add(parseChallenge(challenge));
+                }
+            } catch (ResourceAuthenticationChallengeException e) {
+                throw e;
+            } catch (Exception e) {
                 throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
             }
+
+            return challenges;
+        }
+
+        /**
+         * For multiple challenges in a WWW-Authenticate header value, separate them into multiple
+         * Strings for parsing.
+         *
+         * @param challenges The challenge values to parse.
+         * @return A List of separated challenges.
+         */
+        private static List<String> separateChallenges(final String challenges) {
+            // Split the supplied String on those commas which are not constrained by quotes
+            String[] splitOnUnquotedCommas = challenges.split(REGEX_SPLIT_UNQUOTED_COMMA, -1);
+            sanitizeWhitespace(splitOnUnquotedCommas);
+            List<String> tokensContainingScheme = extractTokensContainingScheme(splitOnUnquotedCommas);
+
+            // init an array to store the out-values
+            String[] outStrings = new String[tokensContainingScheme.size()];
+            for (int ii = 0; ii < outStrings.length; ii++) {
```

Comment:
just curious: ii vs i vs index?

Reply:
Yep -- I tend to favor using `ii` for indexes as `i` isn't very `ctrl-f` searchable. I'm fine with `index` too, mostly `ii` is a matter of habit/style at this point

Reply:
ok.